src/computability/list_blank.lean

```diff
@@ -0,0 +1,563 @@
+  {p : list_blank Γ → Prop} (q : list_blank Γ)
+  (h : ∀ a, p (list_blank.mk a)) : p q := quotient.induction_on' q h
+
+/-- The head of a `list_blank` is well defined. -/
+def list_blank.head {Γ} [inhabited Γ] (l : list_blank Γ) : Γ := list_val.head l
+
+@[simp] theorem list_blank.head_mk {Γ} [inhabited Γ] (l : list Γ) :
+  list_blank.head (list_blank.mk l) = l.head :=
+begin
+  cases l,
+  { simp only [list_blank.mk, list.head],
+    refl, },
+  { simp only [list.head_cons],
+    refl, },
+end
+
+/-- The tail of a `list_blank` is well defined (up to the tail of blanks). -/
+def list_blank.tail {Γ} [inhabited Γ] (l : list_blank Γ) : list_blank Γ := list_val.tail l
+
+@[simp] theorem list_blank.tail_mk {Γ} [inhabited Γ] (l : list Γ) :
+  list_blank.tail (list_blank.mk l) = list_blank.mk l.tail := rfl
+
+/-- We can cons an element onto a `list_blank`. -/
+def list_blank.cons {Γ} [inhabited Γ] (a : Γ) (l : list_blank Γ) : list_blank Γ := list_val.cons a l
+
+@[simp] theorem list_blank.cons_mk {Γ} [inhabited Γ] (a : Γ) (l : list Γ) :
+  list_blank.cons a (list_blank.mk l) = list_blank.mk (a :: l) := rfl
+
+@[simp] theorem list_blank.head_cons {Γ} [inhabited Γ] (a : Γ) :
+  ∀ (l : list_blank Γ), (l.cons a).head = a :=
+quotient.ind' $ by exact λ l, rfl
+
+@[simp] theorem list_blank.tail_cons {Γ} [inhabited Γ] (a : Γ) :
+  ∀ (l : list_blank Γ), (l.cons a).tail = l :=
+quotient.ind' $ by exact λ l, rfl
+
+/-- The `cons` and `head`/`tail` functions are mutually inverse, unlike in the case of `list` where
+this only holds for nonempty lists. -/
+@[simp] theorem list_blank.cons_head_tail {Γ} [inhabited Γ] :
+  ∀ (l : list_blank Γ), l.tail.cons l.head = l :=
+quotient.ind' begin
+  refine (λ l, quotient.sound' (or.inr _)),
+  cases l, {exact ⟨1, rfl⟩}, {refl},
+end
+
+/-- The `cons` and `head`/`tail` functions are mutually inverse, unlike in the case of `list` where
+this only holds for nonempty lists. -/
+theorem list_blank.exists_cons {Γ} [inhabited Γ] (l : list_blank Γ) :
+  ∃ a l', l = list_blank.cons a l' :=
+⟨_, _, (list_blank.cons_head_tail _).symm⟩
+
+/-- The n-th element of a `list_blank` is well defined for all `n : ℕ`, unlike in a `list`. -/
+def list_blank.nth {Γ} [inhabited Γ] (l : list_blank Γ) (n : ℕ) : Γ := list_val.nth l n
+
+@[simp] theorem list_blank.nth_mk {Γ} [inhabited Γ] (l : list Γ) (n : ℕ) :
+  (list_blank.mk l).nth n = l.inth n := rfl
+
+@[simp] theorem list_blank.nth_zero {Γ} [inhabited Γ] (l : list_blank Γ) : l.nth 0 = l.head :=
+list_val.nth_zero l
+
+@[simp] theorem list_blank.nth_succ {Γ} [inhabited Γ] (l : list_blank Γ) (n : ℕ) :
+  l.nth (n + 1) = l.tail.nth n :=
+list_val.nth_succ l n
+
+@[ext] theorem list_blank.ext {Γ} [inhabited Γ] {L₁ L₂ : list_blank Γ} :
+  (∀ i, L₁.nth i = L₂.nth i) → L₁ = L₂ :=
+list_val.ext
+
+/-- Apply a function to a value stored at the nth position of the list. -/
+@[simp] def list_blank.modify_nth {Γ} [inhabited Γ] (f : Γ → Γ) : ℕ → list_blank Γ → list_blank Γ :=
+list_val.modify_nth f
```

Comment:
I made this change but turing_machine still seems broken. The proofs break at a `simp only` - i'm not sure how changing the simp set can affect that. How did you determine that this was the issue?

Reply:
Seems like the reason the first proof breaks is because there are lemmas in the turing_machine file that expect list_blanks but are getting list_vals. I might try rewriting everything so that list_val is simp normal form.